The pokemon / pokestop observable collections are now updated so they no longer flicker on the map.

Jark opened the pull request:
Since I'm on a lower-end mobile and in an area where there's loads of pokestops (London) I noticed that I got a bit of lag / flickering whenever the pokestop collection was updated.

This change is to update the items in the observable collections and not completely remove all items and then re-add them all.

Any questions let me know.

ST-Apps commented:
The code looks fine but I can't test it now (and I can't even get into code details) so I have two questions:

* did you try if it works when the new collection has completely different items? Let's say that the update has 5 items while the previous collection had 7, do the extra 2 element disappear correctly?
* Are you sure that you need $ formatted strings? Can't you just use `nameof` which returns a string?

Jark commented:
1. Yeah, I tested my code since yesterday on my phone and I've stepped through the code in debug code as well to verify it works, but thinking about it I think it needs unit tests, I'll write some for the CollectionUtils methods.
2. Eh, no I don't think I need those no, don't know what I was thinking :P Too dependent on Resharper sometimes :) I'll remove them

denaxin commented:
Wow! This is something I've been waiting for!

jakubsuchybio commented:
Great work! However I just have to point out that your approach is really inefficient. But I guess when here we are dealing with single, max double collection digits, its fine. If you need to keep the ordering of destination, then  better approach, I think, would be to sort the new collection, then iterate through destination collection and on every item you would binary search the new collection. Then you would get O(n*log(n)) complexity instead of O(n^2). Just the thought :)

Jark commented:
@ST-Apps I've made the nameof changes and added a unit test project with tests around the CollectionUtils.Update methods.

@jakubsuchybio I agree they could be made faster, but not worth it atm due to the low size as you say. Feel free to play with it though after this is merged back in, I've added some unit tests so it's easier to modify them :)

Migsi commented:
How would it be to keep already loaded PokeStops in memory, so they won't disappear even when getting away from them? I mean, those stops never move which would make buffering an obvious feature.

Also on the original app you can see pokestops and arenas which are more far away, than the farest on PoGo.

jakubsuchybio commented:
That is interesting thought. But when I think about it I can see few problems:
- You can't show or send to render all of them, that would be rly bad performance wise. Therefore you would need a data structure to hold them with fast searching with range and point. They exist, but are hard to implement.
- I am not sure, if pokestops are updated by most used places in real-time. If yes, then caching could induce visual errors of known places, but they wouldn't be complete.
- You would have to somehow differenciate cached/downloaded because you don't know if some more distant is still used or rdy to be used.

ST-Apps commented:
@Migsi I think I'll have a different kind of solution in some time, but we can't cache them because someone may have used one of it and we need to show it with a different color.

@Jark merging as soon as I finish dealing with some boring stuff :)

Migsi commented:
@ST-Apps It makes sense that the server takes care of the cooldown, I didn't think about that. But you could add a separate local counter which takes care of those gotten out of range? Since the accuracy for pokestops far away doesn't need to be that good, you could use a single counter handling all far stops, decrementing their cooldown every x seconds til it's over.
But maybe there are other possibilities regarding the api, since the original app contains this "feature" too.
However, we'll see later what can be done regarding this. For the moment it's just fine how it is :)

ST-Apps commented:
Came here to merge, conflicting again. Please kill me.

Can you fix it so that I can finally merge this huge change?

Jark commented:
Yeah, no problem, know how it gets with volatile projects. Will rebase the changes tomorrow morning (I'm in Gmt+1 currently).

Cheers,

Jark

ST-Apps commented:
Great, hoping that we'll have a working API again by tomorrow or your collections will always look empty :)

Jark commented:
Hey @ST-Apps, I rebased my changes on the latest master, and you should be able to merge it back now :)

Hopefully we'll see a solution soon for the encrypted api so we can see it in action again :)

ST-Apps commented:
Green merge button = good to go! Merging right now, terrific work! :)

ST-Apps commented:
So, I finally had the chance to test it and there are some issues, but this one is the most important: coverters are not called when update is complete. For example, to get the correct Pokestop icon I need to bind to the FortDataWrapper object, but the PropertyChanged events are not fired on the object itself. Can you please find a way to make it update too?

Jark commented:
@ST-Apps If this were regular WPF I would say that we should use an IMultiValueConverter & MultiBinding, but unfortunately UWP doesn't have those.

Easiest solution I can think of is to wrap the values needed by the Poke stop icon converter in a class (maybe something like PositionData) in FortDataWrapper and do a RaisePropertyChanged on that. 

The poke stop converter should then bind successfully.

I'd love to fix this for you, but am away for about 2 weeks and will have no access to a computer during that time.

Hope this helps.

Cheers,

Jark
